twilio-auth-sample/src/main/java/com/twilio/authsample/main/TokenDetailsActivity.java

```diff
@@ -2,54 +2,156 @@
 
 import android.content.Intent;
 import android.os.Bundle;
+import android.support.annotation.NonNull;
+import android.support.design.widget.Snackbar;
 import android.support.v7.app.ActionBar;
 import android.support.v7.app.AppCompatActivity;
 import android.view.MenuItem;
+import android.widget.TextView;
 
 import com.twilio.authenticator.TwilioAuthenticator;
+import com.twilio.authenticator.external.AuthenticatorObserver;
+import com.twilio.authenticator.external.AuthenticatorToken;
+import com.twilio.authenticator.external.TOTP;
+import com.twilio.authenticator.external.TOTPs;
 import com.twilio.authsample.App;
 import com.twilio.authsample.R;
-import com.twilio.authsample.totp.TokenDetailsFragment;
+import com.twilio.authsample.totp.TokenTimer;
+import com.twilio.authsample.ui.views.AuthyTimerView;
+import com.twilio.authsample.utils.MessageHelper;
 
-public class TokenDetailsActivity extends AppCompatActivity {
+public class TokenDetailsActivity extends AppCompatActivity implements TokenTimer.OnTimerListener, AuthenticatorObserver {
 
     static final String EXTRA_APP_ID = "APP_ID";
+    private static final int TICK_INTERVAL_TIME_MILLIS = 50;
+
+    private TwilioAuthenticator twilioAuthenticator;
+    private TokenTimer tokenTimer;
+    private String appId;
+    private String title;
+    private TOTP currentTOTP;
+    private Snackbar.Callback appDeletedCallback = new Snackbar.Callback() {
+        @Override
+        public void onDismissed(Snackbar transientBottomBar, int event) {
+            super.onDismissed(transientBottomBar, event);
+            finish();
+        }
+    };
+
+    // Views
+    private TextView totpView;
+    private AuthyTimerView authyTimerView;
+    private MessageHelper messageHelper;
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_token_details);
 
-        String appId = getIntent().getStringExtra(EXTRA_APP_ID);
-        String title = getIntent().getStringExtra(Intent.EXTRA_TITLE);
+        twilioAuthenticator = ((App) getApplicationContext()).getTwilioAuthenticator();
```

Comment:
Even though it's very common, since this is a sample app, I would try to avoid this approach. Not sure if it looks good.

We should either inject it with Dagger or retrieve it directly from the singleton's `getInstance()`.

Reply:
I don't want to use dagger, it will add more complexity to a sample project, which could confuse developers. At the same time I want to be able to easily mock TwilioAuthenticator when writing tests, this was the approach that I came up with. I'll try to think on other way to do this or if you have any ideas apart from using dagger I'll be glad to implement it.

Reply:
Don't worry. We can leave it like this, but maybe a comment disclaimer acknowledging that `this should best be injected...`